lib/posix-netlink/netlink.c

```diff
@@ -0,0 +1,339 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2025, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <string.h>
+#include <unistd.h>
+
+#include <uk/netlink_protocol.h>
+#include <uk/socket_driver.h>
+#include <uk/print.h>
+
+#define sock2nlctx(s) ((struct nl_ctx *)posix_sock_get_data(s))
+
+extern struct posix_netlink_protocol posix_netlink_protocol_list_start[];
+extern struct posix_netlink_protocol posix_netlink_protocol_list_end[];
+
+static const struct posix_netlink_protocol *nl_driver_find(int protocol)
+{
+	const struct posix_netlink_protocol *drv;
+
+	drv = posix_netlink_protocol_list_start;
+	while (drv != posix_netlink_protocol_list_end) {
+		if (drv->protocol == protocol)
+			return drv;
+		drv++;
+	}
+	return NULL;
+}
+
+static void *nl_create(struct posix_socket_driver *drv,
+		       int family __maybe_unused, int type, int protocol)
+{
+	const struct posix_netlink_protocol *nl_drv;
+	struct nl_ctx *nl_ctx;
+	int err;
+
+	UK_ASSERT(drv);
+	UK_ASSERT(family == AF_NETLINK);
+
+	/* AF_NETLINK sockets must be created with SOCK_RAW or SOCK_DGRAM */
+	if (unlikely(((type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC)) != SOCK_RAW) &&
+		     ((type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC)) != SOCK_DGRAM))) {
+		err = EINVAL;
+		goto err_out;
+	}
+
+	/* Lookup a driver for the requested protocol */
+	nl_drv = nl_driver_find(protocol);
+	if (unlikely(!nl_drv)) {
+		uk_pr_debug("did not find protocol handler for %d\n", protocol);
+		err = EAFNOSUPPORT;
+		goto err_out;
+	}
+
+	nl_ctx = uk_zalloc(drv->allocator, sizeof(*nl_ctx));
+	if (unlikely(!nl_ctx)) {
+		err = ENOMEM;
+		goto err_out;
+	}
+	nl_ctx->a = drv->allocator;
+	nl_ctx->drv = nl_drv;
+#if CONFIG_LIBPOSIX_PROCESS
+	/* FIXME: PID is here like port number and must be unique. It just
+	 *        starts with PID if free.
+	 */
+	nl_ctx->nl_pid = getpid();
+#endif /* CONFIG_LIBPOSIX_PROCESS */
+
+	nl_ctx->nl_recvqueue = uk_mbox_create(nl_ctx->a, 512);
+	if (unlikely(!nl_ctx->nl_recvqueue)) {
+		err = ENOMEM;
+		goto err_free_ctx;
+	}
+
+	/* Optional driver initialization */
+	if (nl_ctx->drv->ops->create) {
+		err = nl_ctx->drv->ops->create(nl_ctx);
+		if (unlikely(err))
+			goto err_free_recvqueue;
+	}
+
+	uk_pr_debug("Created netlink socket: %p, protocol: %d, drv: %s\n",
+		    nl_ctx, nl_ctx->drv->protocol, nl_ctx->drv->libname);
+	return nl_ctx;
+
+err_free_recvqueue:
+	uk_mbox_free(nl_ctx->a, nl_ctx->nl_recvqueue);
+err_free_ctx:
+	uk_free(nl_ctx->a, nl_ctx);
+err_out:
+	return ERR2PTR(-err);
+}
+
+static int nl_close(posix_sock *s)
+{
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+	struct uk_streambuf *nlbuf;
+
```

Comment:
Perhaps it would be worth having a `UK_ASSERT(s)` before deriving the netlink context. Ditto for the rest of ops.